enzyme/Enzyme/EnzymeLogic.cpp

```diff
@@ -2820,7 +2821,9 @@ Function *EnzymeLogic::CreatePrimalAndGradient(
 
     for (size_t i = 0; i < constant_args.size(); ++i) {
       auto arg = constant_args[i];
-      if (arg == DIFFE_TYPE::DUP_ARG) {
+      auto argTypeInfo = oldTypeInfo.Function->args().begin() + i;
+      if (arg == DIFFE_TYPE::DUP_ARG &&
+          !argTypeInfo->getType()->isPointerTy()) {
```

Comment:
Shouldn't all active arguments be dup_arg (so perhaps change that to an assert)? Also please use type analysis info here rather than a pure pointer check.

Reply:
Yes, this should become unnecessary once I've figured out the activity of return check